### PR TITLE
Graphql / Uplift branch_protection resource

### PR DESCRIPTION
**Pull request by patrickmarabeas**
Uplifts the branch_protection resource to use v4 API.

API surface has changed slightly.

Migration function only worries about:
- migration of `id` from repository name to the current branch protection's global object ID
- branch -> pattern

**notes:**
- This branch will be rebased as necessary
- Shared consts and functions will be abstracted int util_* files as required, I've simply included them within the one file for simplicity for the time being
- This constitutes a preliminary migration to the v4 API
- Tests will be written / updated once the API surface is nailed down

### CODE REVIEW

**Comment by patrickmarabeas** [comment]
Unfortunately it seems as though v3 has continued to provide _some_ newer features - see #305. I'll do some testing to see whether conditionally updating branch protection with the v3 API if the branch exists is a feasible solution.

**Comment by twexler** [comment]
Looks like this one kind of died, any chance we could revive it? Could really use the ability to specify patterns for branch protections!

Willing to help if needed.

**Comment by patrickmarabeas** [comment]
@twexler Still alive. There's currently discussion on how to roll v4 resources into this provider. 

If you are wanting something now - have a look at https://github.com/patrickmarabeas/terraform-provider-github-v4 I'm using this as a test bed for v4 resources. Documentation is certainly lacking, but happy to add to it if questions are raised.

**Comment by twexler** [comment]
Not super urgent, just a nice thing to have! I saw the discussion in #83 after I wrote that message.

I'd be happy to help with docs or anything else that is needed on the v4 transition!

**Comment by jcudit** [comment]
@patrickmarabeas @twexler -

I'm hoping to cut a [new release](https://github.com/terraform-providers/terraform-provider-github/milestone/4) with the [incremental GraphQL changes accumulated so far](https://github.com/terraform-providers/terraform-provider-github/milestone/4?closed=1).

What are thoughts around deferring this uplift to another release? I am unclear on the confidence of what is outlined in this PR, but happy to resolve conflicts and merge if this approach is still favoured.

**Comment by patrickmarabeas** [comment]
Yes, further work is required on this PR.

**Comment by wuurrd** [comment]
:+1: this PR is extremely useful to fill in the gaps of missing v3 API stuff

**Comment by patrickmarabeas** [comment]
Apologies, haven't had the time recently to focus on getting this squared away.

**Comment by patrickmarabeas** [comment]
Uplifting the tests at the moment, but it would be good to have a review on the changes to the parameters and what not.

There is no available v3 API endpoints to add the missing functionality (`required_linear_history`, `allow_force_pushes`, `allow_deletions `), that doesn't also require you to provide ALL the fields (https://developer.github.com/v3/repos/branches/#update-branch-protection) which simply doesn't align with v4 (`actor_ids` vs `apps`, `teams`, `users` for instance).

CC @jcudit @anGie44

**Comment by patrickmarabeas** [comment]
I've pulled the uplifted resource into its own files to avoid merge hell, and commented out the legacy files. This can be corrected when it's ready to be merged.

**Comment by jcudit** [comment]
🥇 for patience on seeing this one through.  I'll merge this shortly and plan to follow on with a few commits to:
- update the tests to run with the latest test suite
- rename the `branch_protection_v4` resources back to `branch_protection`

**Review by jcudit — APPROVED** [review]

**Comment by patrickmarabeas** [comment]
Awesome, great to have this in!

Just an FYI for the future, you can squash `fixup!` commits by way of:  `git rebase --autosquash --interactive master`.

**Comment by Xorima** [comment]
Why is this change not in the release notes as breaking change or updated in the docs?

**Comment by lsowen** [comment]
@patrickmarabeas in v3, you could specify `apps` for push restrictions.  It doesn't look like it is possible to get the `node_id` for an app for use with v4?

**Comment by Krenair** [comment]
Yeah, this being listed as an enhancement for 3.1.0 instead of a breaking change seems like a mistake?

File: github/util_v4_branch_protection.go [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+}
+
+func setActorIDs(actors []ActorTypes) []string {
+	pushActors := make([]string, 0, len(actors))
+	for _, a := range actors {
+		if a.Actor.Team != (Actor{}) {
+			pushActors = append(pushActors, a.Actor.Team.ID.(string))
+		}
+		if a.Actor.User != (Actor{}) {
+			pushActors = append(pushActors, a.Actor.Team.ID.(string))
```
Comment (usmonster) [inline github/util_v4_branch_protection.go:168]: Is this a bug? Should it instead be:
```suggestion
			pushActors = append(pushActors, a.Actor.User.ID.(string))
```
?

Comment (jarylc) [reply github/util_v4_branch_protection.go:168]: Will be rectified in https://github.com/integrations/terraform-provider-github/pull/819

Comment (usmonster) [reply github/util_v4_branch_protection.go:168]: I've just seen your comment—there's a more complete fix (since the copy-paste error propagated) in #824.

**Pull request merged**